File: Algorithms/0155_Min_Stack/Python/Min_Stack_Solution_1.py

# Space: O(n)
# Time: O(n)

class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.data = []

    def push(self, x: int) -> None:
        if len(self.data) == 0:
            self.data.append((x, x))
        else:
            self.data.append((x, min(x, self.data[-1][1])))

    def pop(self) -> None:
        self.data.pop()

    def top(self) -> int:
        return self.data[-1][0]

    def getMin(self) -> int:
        return self.data[-1][1]

# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: Algorithms/0155_Min_Stack/Java/MinStackSolution1.java

// Space: O(n)
// Time: O(n)

class MinStack {
    private Stack<int[]> stack;
    private int min = Integer.MAX_VALUE;

    public MinStack() {
        stack = new Stack<>();
    }

    public void push(int val) {
        min = Math.min(min, val);
        int[] temp = new int[]{val, min};
        stack.add(temp);
    }

    public void pop() {
        stack.pop();
        min = stack.isEmpty()? Integer.MAX_VALUE:getTop()[1];
    }

    public int top() {
        return getTop()[0];
    }

    public int getMin() {
        return getTop()[1];
    }

    private int[] getTop() {
        return stack.peek();
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */




